[FEA] Implement kfold.

trivialfis opened the pull request:
Replaces https://github.com/rapidsai/cuml/pull/5940 .

Closes https://github.com/rapidsai/cuml/issues/7088

betatim commented on python/cuml/tests/test_stratified_kfold.py:
```diff
@@ -63,3 +65,53 @@ def test_invalid_folds(n_splits):
+@pytest.mark.parametrize("shuffle", [True, False])
+@pytest.mark.parametrize("n_splits", [5, 10])
+@pytest.mark.parametrize(
+    "rs",
+    [
+        1,
+        np.random.RandomState(1),
+        cp.random.RandomState(1),
+        np.random.default_rng(1),
+        cp.random.default_rng(1),
+    ],
+)
+def test_kfold(shuffle: bool, n_splits: int, rs) -> None:
```
```suggestion
    "random_state",
    [
        1,
        np.random.RandomState(1),
        cp.random.RandomState(1),
        np.random.default_rng(1),
        cp.random.default_rng(1),
    ],
)
def test_kfold(shuffle, n_splits, random_state) -> None:
```

I'd spell out random state to make it easier to read, remove the type annotations

trivialfis commented on python/cuml/cuml/model_selection/_split.py:
```diff
@@ -419,7 +447,90 @@ def _process_df_objs(
+            n_splits=n_splits, shuffle=shuffle, random_state=random_state
+        )
+
+    def split(self, x, y=None) -> Generator[Tuple[Any, Any], None, None]:
```
Added to the base class, with `groups` removed as it's not supported by any existing kfold.

trivialfis commented on python/cuml/cuml/model_selection/_split.py:
```diff
@@ -419,7 +447,90 @@ def _process_df_objs(
+    shuffle :
+        Whether to shuffle the samples before splitting.
+    random_state : int/Generator/RandomState/None (default=None)
+        Random seed. Ignored if shuffle is False.
```
Thank you for sharing. That makes sense. I updated the doc string as you suggested.

trivialfis commented:
> I think we can consider it for 25.10, but I'd like the comments by @betatim to be addressed.

Thank you. Kept the targeted branch unchanged.

@betatim Thank you for the detailed comments. I have addressed most of the issues. I kept the existing train_test_split and stratified kfold unchanged. But in the future, they might need some cleanup to meet the new cuml conventions.

csadorf commented on python/cuml/cuml/model_selection/_split.py:
```diff
@@ -419,7 +421,106 @@ def _process_df_objs(
+
+    """
+
+    def __init__(self, n_splits=5, shuffle=False, random_state=None):
```
We should be using keyword-only arguments after `n_splits`.

```suggestion
    def __init__(self, n_splits=5, * shuffle=False, random_state=None):
```

This is a general rule, but also matches the sklearn API.

csadorf commented on python/cuml/cuml/model_selection/_split.py:
```diff
@@ -419,7 +421,106 @@ def _process_df_objs(
+    Each fold is then used once as a validation set while the k - 1 remaining folds form
+    the training set.
+
+    Parameters
```
We should document default values for all parameters.

csadorf commented on python/cuml/cuml/model_selection/_split.py:
```diff
@@ -419,7 +421,106 @@ def _process_df_objs(
+
+        Yields
+        ------
+        train_idx : CuPy ndarray
+            The training set indices for that split.
+
+        test_idx : CuPy ndarray
+            The testing set indices for that split.
```
```suggestion
        train : CuPy ndarray
            The training set indices for that split.

        test : CuPy ndarray
            The testing set indices for that split.
```

csadorf commented on python/cuml/cuml/model_selection/_split.py:
```diff
@@ -419,7 +421,106 @@ def _process_df_objs(
+        raise NotImplementedError()
+
+
+class KFold(_KFoldBase):
```
Why are we not supporting the `get_n_splits` method?

trivialfis replied:
added

betatim commented:
> @betatim Thank you for the detailed comments. I have addressed most of the issues. I kept the existing train_test_split and stratified kfold unchanged. But in the future, they might need some cleanup to meet the new cuml conventions.

Thanks for being efficient and nice with addressing them! I think it is a good idea to leave clean up/making things more uniform to a separate PR. Having a PR that changes/adds new things and does clean up is trickier to review because you need to double check if this is just a "moved things around" change or a new addition.

What do you have in mind with "the new cuml conventions"?

betatim commented on python/cuml/cuml/model_selection/_split.py:
```diff
@@ -419,7 +421,127 @@ def _process_df_objs(
+
+    @abstractmethod
+    def split(self, x, y):
+        """Generate indices to split data into training and test set.
```
I think we need to duplicate/move this to the public class(es) if we want it to show up nicely. At least I don't think that docstrings are "inherited"

csadorf replied:
I think they are "inherited."

```
cuml.model_selection.KFold.split = split(self, x, y=None)
    Generate indices to split data into training and test set.

    Parameters
    ----------
    X : array-like of shape (n_samples, n_features)
        Training data, where `n_samples` is the number of samples
        and `n_features` is the number of features.

    y : array-like of shape (n_samples,), default=None
        The target variable for supervised learning problems.

    Yields
    ------
    train : CuPy ndarray
        The training set indices for that split.

    test : CuPy ndarray
        The testing set indices for that split.
```

trivialfis replied:
I don't know how to build doc for cuML yet. But we use inherited members for sphinx build I XGBoost. Makes life easier.

csadorf commented:
Please re-open this for branch-25.12.
